### PR TITLE
IGNITE-17843 Use partitionless pageId in BplusInnerIo

### PR DESCRIPTION
**Pull request by tkalkirill**
https://issues.apache.org/jira/browse/IGNITE-17843

### CODE REVIEW

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java [hunk 1]
```diff
@@ -100,9 +100,9 @@
  * <p>All pages in the tree are divided into levels. Leaves are always at the level {@code 0}. Levels of inner pages are thus positive.
  * Each
  * level represents a singly linked list - each page has a link to the <i>forward</i> page at the same level. It can be retrieved by calling
- * {@link BplusIo#getForward(long)}. This link must be a zero if there's no forward page. Forward links on level {@code 0} allow iterating
- * tree's keys and values effectively without traversing any inner nodes ({@code AbstractForwardCursor}). Forward links in inner nodes have
- * different purpose, more on that later.
+ * {@link BplusIo#getForwardPageId(long, int)}. This link must be a zero if there's no forward page. Forward links on level {@code 0} allow
```
Comment (ibessonov) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java:103]: Old name was good enough, why did you rename it? Please change it back.

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java [hunk 2]
```diff
@@ -113,8 +113,8 @@
  * </code></pre>
  * There are {@code N} items and {@code N+1} links. Each link points to page of a lower level. For example, pages on level {@code 2} always
  * point to pages of level {@code 1}. For an item {@code i} left subtree is defined by {@code link(i)} and right subtree is defined by
- * {@code link(i+1)} ({@link BplusInnerIo#getLeft(long, int)} and {@link BplusInnerIo#getRight(long, int)}). All items in the left subtree
- * are less or equal to the original item (basic property for the trees).
+ * {@code link(i+1)} ({@link BplusInnerIo#getLeftPageId(long, int, int)} and {@link BplusInnerIo#getRightPageId(long, int, int)}).
```
Comment (ibessonov) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java:116]: Same for these two. Please also return the old formatting back after renaming, if it fits

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/util/PartitionlessLinks.java [hunk 3]
```diff
@@ -92,4 +100,45 @@ public static void writeToBuffer(ByteBuffer buffer, long link) {
...
+
+    /**
+     * Reads a partitionless page ID from the memory.
+     *
+     * @param addr Address in memory.
+     * @param off Offset from {@code addr} in bytes.
+     * @param partId Partition ID.
+     * @return Page ID with partition ID.
+     */
+    public static long readPartitionlessPageId(long addr, int off, int partId) {
```
Comment (ibessonov) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/util/PartitionlessLinks.java:112]: What's the difference between this and `readPartitionlessLink`?
There should be no difference. Not only this implementation is buggy, it's also a copy-paste.
Please make a single method and update comment in it. Renaming here would be fine

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/util/PartitionlessLinks.java [hunk 4]
```diff
@@ -92,4 +100,45 @@ public static void writeToBuffer(ByteBuffer buffer, long link) {
...
+     *     <li>{@link PageIdUtils#flag Flag} - 1 byte;</li>
+     *     <li>{@link PageIdUtils#pageIndex Page index} - 4 bytes.</li>
+     * </ul>
+     *
+     * @param addr Address in memory.
+     * @param off Offset from {@code addr} in bytes.
+     * @param pageId Page ID.
+     * @return Number of bytes written (equal to {@link #PARTITIONLESS_PAGE_ID_SIZE_BYTES}).
+     */
+    public static long writePartitionlessPageId(long addr, int off, long pageId) {
```
Comment (ibessonov) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/util/PartitionlessLinks.java:135]: Remove this method as well

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java (hunk 1 above)
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java:103]: Fix it

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java (hunk 2 above)
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/BplusTree.java:116]: Fix it

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/util/PartitionlessLinks.java (hunk 3 above)
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/util/PartitionlessLinks.java:112]: Fix it

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/io/BplusInnerIo.java [hunk 5]
```diff
@@ -81,23 +84,24 @@ public final long getLeft(long pageAddr, int idx) {
     public final void setLeft(long pageAddr, int idx, long pageId) {
         assertPageType(pageAddr);
 
-        putLong(pageAddr, offset0(idx, SHIFT_LEFT), pageId);
+        writePartitionless(pageAddr + offset0(idx, SHIFT_LEFT), link(pageId, 0));
```
Comment (ibessonov) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/io/BplusInnerIo.java]: I believe that having a `link(pageId, 0)` is wrong here.

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/io/BplusInnerIo.java [hunk 6]
```diff
@@ -81,23 +84,24 @@ public final long getLeft(long pageAddr, int idx) {
...
 
-        assert pageId == getLeft(pageAddr, idx);
+        assert pageId == getLeft(pageAddr, idx, partitionId(pageId));
     }
 
     /**
-     * Returns right page id for item.
+     * Returns right page ID for item.
```
Comment (ibessonov) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/io/BplusInnerIo.java:93]: Please don't capitalize it. First of all, why would you touch a comment that has nothing to do with your changes. Second, "id" is grammatically correct.

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/io/BplusInnerIo.java [hunk 7]
```diff
@@ -81,23 +84,24 @@ public final long getLeft(long pageAddr, int idx) {
...
      */
-    public final long getRight(long pageAddr, int idx) {
-        return getLong(pageAddr, offset0(idx, shiftRight));
+    public final long getRight(long pageAddr, int idx, int partId) {
+        return readPartitionless(partId, pageAddr, offset0(idx, shiftRight));
     }
 
     /**
-     * Returns right page id for item.
+     * Sets right page ID for item.
```
Comment (ibessonov) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/io/BplusInnerIo.java:104]: Same here, please leave "id"

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/io/BplusIo.java [hunk 8]
```diff
@@ -346,14 +357,15 @@ public abstract void copyItems(
...
+            boolean needRowBytes,
+            int partId
     ) throws IgniteInternalCheckedException {
         assertPageType(pageAddr);
 
         int cnt = getCount(pageAddr);
 
         // Move right all the greater elements to make a free slot for a new row link.
-        copyItems(pageAddr, pageAddr, idx, idx + 1, cnt - idx, false);
+        copyItems(pageAddr, pageAddr, idx, idx + 1, cnt - idx, false, partId);
```
Comment (ibessonov) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/io/BplusIo.java:368]: `partId` doesn't seem necessary for copying, why is it here?

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/util/PartitionlessLinks.java [hunk 9]
```diff
@@ -29,25 +29,26 @@
...
+     * Reads a partitionless link or page ID from the memory.
      *
      * @param partitionId Partition ID.
      * @param pageAddr Page address.
      * @param offset Data offset.
-     * @return Partitionless link.
+     * @return Link or page ID with partition ID.
      */
-    public static long readPartitionlessLink(int partitionId, long pageAddr, int offset) {
+    public static long readPartitionless(int partitionId, long pageAddr, int offset) {
```
Comment (ibessonov) [inline modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/util/PartitionlessLinks.java:51]: I also asked to update comment in this method. Line `Page with index 0 is never a data page.` makes no sense now, it should be changed

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/io/BplusInnerIo.java (hunk 5 above)
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/io/BplusInnerIo.java]: Fix it

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/io/BplusInnerIo.java (hunk 6 above)
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/io/BplusInnerIo.java:93]: Fix it

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/util/PartitionlessLinks.java (hunk 9 above)
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/util/PartitionlessLinks.java:51]: fix it

File: modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/io/BplusIo.java (hunk 8 above)
Comment (tkalkirill) [reply modules/page-memory/src/main/java/org/apache/ignite/internal/pagememory/tree/io/BplusIo.java]: Fix it

**Review by ibessonov — APPROVED** [review]
